prusti-interface/src/specs/typed.rs

```diff
@@ -134,7 +136,67 @@ impl SpecificationItem<bool> {
     }
 }
 
+#[derive(Debug)]
+pub enum ProcedureSpecificationKindError {
+    /// Occurs whenever the relation between [ProcedureSpecificationKind]s is violated.
+    /// The relation is: predicates ⊂ pure functions ⊂ impure functions
+    /// This validation is not automatically performed during refinement. It needs to be invoked
+    /// manually with [SpecificationItem<ProcedureSpecificationKind>::validate]
+    InvalidSpecKindRefinement(ProcedureSpecificationKind, ProcedureSpecificationKind)
+}
+
+impl SpecificationItem<ProcedureSpecificationKind> {
+    pub fn is_pure(&self) -> Result<bool, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(matches!(self.extract_with_selective_replacement(),
+                Some(ProcedureSpecificationKind::Pure) | Some(ProcedureSpecificationKind::Predicate(_))))
+    }
+
+    pub fn is_impure(&self) -> Result<bool, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(match self.extract_with_selective_replacement() {
+            Some(refined) => matches!(refined, ProcedureSpecificationKind::Impure),
+            _ => true,
+        })
+    }
+
+    pub fn get_predicate_body(&self) -> Result<Option<&LocalDefId>, ProcedureSpecificationKindError> {
+        self.validate()?;
+
+        Ok(match self.extract_with_selective_replacement() {
+            Some(ProcedureSpecificationKind::Predicate(pred_id)) => Some(pred_id),
+            _ => None
+        })
+    }
+
+    /// Ensures that refined [ProcedureSpecificationKind]s are valid.
+    /// See [ProcedureSpecificationKindError] for detailed error descriptions.
+    pub fn validate(&self) -> Result<(), ProcedureSpecificationKindError> {
+        if let SpecificationItem::Refined(base, refined) = self {
+            match (base, refined) {
+                (ProcedureSpecificationKind::Predicate(_), ProcedureSpecificationKind::Pure) |
+                (ProcedureSpecificationKind::Predicate(_), ProcedureSpecificationKind::Impure) |
+                (ProcedureSpecificationKind::Pure, ProcedureSpecificationKind::Impure) =>
+                    Err(ProcedureSpecificationKindError::InvalidSpecKindRefinement(*base, *refined)),
+                _ => Ok(())
```

Comment:
Let's flip this match around: match on kind refinements that are valid, and make the `_` case raise the error. This should be a little bit more robust against future extensions.

Reply:
Done